### PR TITLE
Using sqrt for better precision in cosine_similarity

### PR DESCRIPTION
**Pull request by ailzhang**
address comment in #18168 .
Testing in CI...

### CODE REVIEW

**Review by apaszke — COMMENTED** [review]
Sorry, why? It doesn't seem to change anything, except that it moves the clamp operation wrt. the inversion operation, and it's actually likely to be slower?

**Comment by ailzhang** [comment]
@apaszke I'm not entirely sure about it, here is the comment from @colesbury in https://github.com/pytorch/pytorch/pull/18168#discussion_r267450164.

**Comment by colesbury** [comment]
@apaszke it's more accurate (and it's also not any slower). rsqrt(x) is implemented as 1/sqrt(x) on CPU. The whole point of the change was to prevent cosine_similarity from returning values larger than one. Using rsqrt defeated that goal.

With float32:
`x / sqrt(x * x)` is always `1.0` if `x>0` and `x*x` is normal

That's no longer true with rsqrt. It breaks for ~15% of the possible values.

This holds for both CPU and CUDA (at least on recent GPUs).

**Comment by colesbury** [comment]
@ailzhang, I think #18168 also caused NaN gradients when the product of the magnitudes is zero.

```python
import torch
x = torch.randn(10).requires_grad_()
y = torch.zeros_like(x).requires_grad_()
torch.cosine_similarity(x, y, 0).sum().backward()
print(x.grad)  # should be zero
print(y.grad)  # large with eps>0 
```

Please add a regression test for this.  

cc @zasdfgbnm

File: aten/src/ATen/native/Distance.cpp [hunk 1]
```diff
@@ -88,8 +88,8 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   Tensor w12 = at::sum(x1 * x2, dim);
   Tensor w1 = at::sum(x1 * x1, dim);
   Tensor w2 = at::sum(x2 * x2, dim);
-  Tensor n12 = (w1 * w2).rsqrt_().clamp_max(1.0 / eps);
-  return w12.mul_(n12);
+  Tensor n12 = (w1 * w2).sqrt_().clamp_min(eps);
```
Comment (colesbury) [inline aten/src/ATen/native/Distance.cpp]: my guess is that you'll need to perform the clamp_min before the sqrt() to avoid NaNs in the derivative

**Comment by zasdfgbnm** [comment]
@ailzhang @colesbury Thanks for the super fast fix!!

File: aten/src/ATen/native/Distance.cpp (hunk 1 above)
Comment (ailzhang) [reply aten/src/ATen/native/Distance.cpp]: Yea I can do that, just double check we also need to update doc for it? As here https://github.com/pytorch/pytorch/blob/670f509984efce8aab0b0b5b166ff01472af6f1c/torch/nn/functional.py#L2838 we seem to imply we take `clamp_min` after `sqrt`.

Comment (ailzhang) [reply aten/src/ATen/native/Distance.cpp]: nvm, please ignore my naive question :P

**Review by colesbury — APPROVED** [review]

**Comment by apaszke** [comment]
@colesbury So you're saying that
```
x * (1 / sqrt(x * x))
```
is what's causing the issues, compared to this which this patch uses:
```
x / sqrt(x * x)
```

I thought the whole problem was in computing the sqrt twice?

**Comment by colesbury** [comment]
@apaszke both formulations `x / (sqrt(x) * sqrt(x))` and `x * (1 / sqrt(x * x))` cause precision issues. The comment https://github.com/pytorch/pytorch/issues/18057#issuecomment-474142662 mentions the first formulation because the code at the time reduced to that expression when the vectors `x1` and `x2` are equal. The comment was not intended to suggest that all formulations that compute sqrt once are precise.

The PR #18168 had a two other issues: NaN gradients when one vector was zero and a floating-point division by zero when eps=0.
